### ISSUE TITLE
Add jax.random.choice

### ISSUE THREAD

**Issue by Thenerdstation** [google/jax#2070]
It would be nice to have a random weighted choice in jax. We need this to sample bitstrings from a quantum wave function.

**Comment by jacobjinkelly** [comment]
I ran into the same issue, I got around it by implementing the function described in this [blog post](https://timvieira.github.io/blog/post/2019/09/16/algorithms-for-sampling-without-replacement/). Ideally `jax.random.choice` would be implemented, but I suppose it's tricky to get it to match the API for `np.random.choice`. Perhaps `jax.random` could have a simpler sampling without replacement function?

**Comment by jekbradbury** [comment]
This is a significant missing feature. I'm curious @Thenerdstation if you also need sampling without replacement or if you're happy with a single sample/sampling with replacement (both much easier).

**Comment by Thenerdstation** [comment]
Sampling with replacement is what we need for our use case

**Comment by grisaitis** [comment]
seconded

should this take a probability vector `p`? `categorical` takes a `logits` vector. (was this for performance reasons?)

guessing api would be something like 
```
jax.random.choice(key, x, size=None, replace=True, p=None, axis=0)
```

**Comment by grisaitis** [comment]
potentially relevant: `permutation` was requested https://github.com/google/jax/issues/1526 with PR pending https://github.com/google/jax/pull/1568

**Comment by mtthss** [comment]
I also have been in need of this several times.
Both the with and without replacement case.

**Comment by jakevdp** [comment]
I wrote an implementation of unweighted `random.choice` in #3463

Weighted will take a bit more thought.

**Comment by jacobjinkelly** [comment]
@jakevdp Not sure if this is the best way, but based off this [blog post](https://timvieira.github.io/blog/post/2019/09/16/algorithms-for-sampling-without-replacement/) I implemented weighted sampling without replacement (`swor`) [here](https://github.com/jessebett/jax/blob/hk-nodes/examples/timeseries.py#L711-L719)

**Comment by jakevdp** [comment]
Thanks @jacobjinkelly – that's really cool!

**Comment by mtthss** [comment]
Can we merge the two solutions? so that we have choice with and without replacement and both weighted and unweighted?

**Comment by jakevdp** [comment]
Yes, that’s my plan.

**Comment by jakevdp** [comment]
OK, #3463 now implements `random.choice` with or without replacement, and with or without weights.
